src/app/icd/ICDManager.cpp

```diff
@@ -228,5 +254,74 @@ void ICDManager::OnTransitionToIdle(System::Layer * aLayer, void * appState)
     pIcdManager->mStateObserver->OnTransitionToIdle();
 }
 
+bool ICDManager::EnsureChipStackLock()
+{
+    // TODO implement platform agnostic lock creation/mechanism that can be used for ICD without locking the chip stack
+    bool requireLock = false;
+
+#if CHIP_STACK_LOCK_TRACKING_ENABLED
+    requireLock = !DeviceLayer::PlatformMgr().IsChipStackLockedByCurrentThread();
+#endif
+
+    if (requireLock)
+    {
+        DeviceLayer::PlatformMgr().LockChipStack();
+    }
+    return requireLock;
+}
+
+/* ICDNotify observer functions. */
+void ICDManager::KeepActiveRequest(KeepActiveFlags request, bool set)
+{
+    bool requiresUnlock = this->EnsureChipStackLock();
+
+    if (request == KeepActiveFlags::kExchangeContextOpen)
+    {
+        // There can be multiple open exchange context at the same time.
+        // Keep track of the requests count.
+        if (set)
+        {
+            this->OpenExchangeContextCount++;
+            this->SetKeepActiveModeRequirements(request, set);
+        }
+        else
+        {
+            if (this->OpenExchangeContextCount > 0)
+            {
+                this->OpenExchangeContextCount--;
+            }
+            else
+            {
+                ChipLogError(DeviceLayer, "The ICD Manager did not account for ExchangeContext closure");
+            }
+
+            if (this->OpenExchangeContextCount == 0)
+            {
+                this->SetKeepActiveModeRequirements(request, set);
+            }
+        }
+    }
+    else /* kCommissioningWindowOpen, kFailSafeArmed */
```

Comment:
the comment should provably say `!KeepActiveFlags::kExchangeContextOpen` since that is the actual branch meaning. We could give commissioningwindowopen and failsafearmed as examples after that.